File: Problems/No.1976_Number_of_Ways_to_Arrive_at_Destination/number_of_ways_to_arrive_at_destination.py
class Solution(object):
    def countPaths(self, n, roads):
        """
        :type n: int
        :type roads: List[List[int]]
        :rtype: int
        """

        # global parameters
        MOD = int(1e9 + 7)
        edges = [[] for _ in range(n)]
        
        # start computing
        # construct graph (edge relationships)
        for [node1, node2, dis] in roads:
            edges[node1].append([node2, dis])
            edges[node2].append([node1, dis])

        # traverse using Dijkstra's Algorithm
        # local parameters
        traverse_count = 0
        min_path = []
        dist, paths = [float('inf')] * n, [0] * n

        # start traverse
        dist[0] = 0
        paths[0] = 1
        heapq.heappush(min_path, [0, 0])

        while (traverse_count < n):
            [curr_dist, curr_node] = heapq.heappop(min_path)
            if (dist[curr_node] < curr_dist):
                continue
            else:
                traverse_count += 1
                # update the distance to the adjacent node from current node
                for [adj_node, edgeW] in edges[curr_node]:
                    updated_dist = curr_dist + edgeW
                    
                    if (updated_dist < dist[adj_node]):
                        dist[adj_node] = updated_dist
                        heapq.heappush(min_path, [updated_dist, adj_node])
                        paths[adj_node] = paths[curr_node]
                    elif (updated_dist == dist[adj_node]):
                        paths[adj_node] = (paths[adj_node] + paths[curr_node]) % MOD

        return (paths[-1] % MOD)

File: Problems/No.1976_Number_of_Ways_to_Arrive_at_Destination/number_of_ways_to_arrive_at_destination.cpp
class Solution {
public:
    int countPaths(int n, vector<vector<int>>& roads) {
        
        // global parameters
        static const int mod = 1e9 + 7;
        std::vector<std::pair<long long, long long>> edges[n];

        // start computing
        // construct graph (edge relationships)
        for (auto &road: roads) {

            edges[road[0]].push_back({road[1], road[2]});
            edges[road[1]].push_back({road[0], road[2]});

        }

        // traverse using Dijkstra's Algorithm
        // local parameters
        int traverse_count = 0;
        std::priority_queue<
            std::pair<long long, long long>, 
            std::vector<std::pair<long long, long long>>, 
            std::greater<>> min_path;
        std::vector<long long> dist(n, LLONG_MAX), paths(n, 0);

        // start traverse
        dist[0] = 0;
        paths[0] = 1;
        min_path.push({0, 0});
        while (traverse_count < n) {

            auto [curr_dist, curr_node] = min_path.top();
            min_path.pop();

            if (dist[curr_node] < curr_dist) continue;
            else {

                // update the distance to the adjacent node from current node
                for (auto &[adj_node, edgeW]: edges[curr_node]) {

                    auto updated_dist = curr_dist + edgeW;
                    if (updated_dist < dist[adj_node]) {

                        dist[adj_node] = updated_dist;
                        min_path.push({updated_dist, adj_node});
                        paths[adj_node] = paths[curr_node];

                    }
                    else if (updated_dist == dist[adj_node]) 
                        paths[adj_node] = (paths[adj_node] + paths[curr_node]) % mod;

                }
                traverse_count++;

            }

        }

        return (paths[n-1] % mod);

    }
};
